lib/tree.rb

```diff
@@ -1,41 +1,93 @@
 class NoApplesError < StandardError; end
+class DeadTreeError < StandardError; end
 
 class Tree
-  attr_#fill_in :height, :age, :apples, :alive
 
-  def initialize
+  attr_reader :height, :age, :apples
+  attr_accessor :alive
+
+  def initialize(height: 0, age: 0, apples: [], alive: true, rand_seed: nil)
+    @height = height
+    @age = age
+    @apples = apples
+    @alive = alive
+    @prng = Random.new(rand_seed || Random.new_seed)
   end
 
   def age!
+    raise DeadTreeError, "This tree is dead" if self.dead?
+    @age += 1
+
+
+    # Simple check if a tree has become too old.
+    if @age >= 10 && age + @prng.rand(0..10) > 20
+      @alive = false
+    end
+
+    # What to do when a tree is alive and is aging
+    if @alive
+      inc_height!
+      add_apples
+    end
+  end
+
+  def inc_height!
+    @height += @prng.rand(1..3)
   end
 
   def add_apples
+    raise DeadTreeError, "This tree is dead" if self.dead?
+
+    count = @height * @prng.rand(1..3)
+    color = Apple.colors.sample
+    diameter = @prng.rand(4..8)
+
+    count.times { @apples.push(Apple.new(color, diameter)) }
   end
 
   def any_apples?
+    @apples.size > 0
   end
 
   def pick_an_apple!
     raise NoApplesError, "This tree has no apples" unless self.any_apples?
+    @apples.shift
   end
 
   def dead?
+    @alive == false
   end
 end
 
+
 class Fruit
+  attr_reader :has_seeds
+
   def initialize
-    has_seeds = true
+    @has_seeds = true
   end
 end
 
-class Apple <
-  attr_reader #what should go here 
+
+class Apple < Fruit
+  attr_reader :color, :diameter
+
+  @@colors = ['red', 'yellow', 'green']
```

Comment:
class variables arent recommended.

https://stackoverflow.com/questions/40819068/why-is-using-a-class-variable-in-ruby-considered-a-code-smell/40819118

instead you might use a constant or encapsulate them in a class method.

```ruby
COLORS = %w(red yellow green).freeze
```

```ruby
  def self.colors
    ['red', 'yellow', 'green']
  end
```

relatedly the cool thing about `class << self` is that you can use it to organize your private class methods in the same way you would handle instance methods.

```ruby

class << self
  def foo
     bar
  end
  
  private

  def bar
    :baz
  end
end
```